### PR TITLE
support calculate relation size by relfilenode

### PR DESCRIPTION
**Pull request by zhrt123**
add unit test for this feature

### CODE REVIEW

File: diskquota_utility.c [hunk 1]
```diff
@@ -1105,3 +1106,58 @@ get_rel_oid_list(void)
 	}
 	return oidlist;
 }
+
+/*
+ * calculate size of (one fork of) a table in transaction
```
Comment (higuoxing) [inline diskquota_utility.c:1111]:
```suggestion
 * calculate size of (all fork of) a table in transaction
```

File: diskquota_utility.c [hunk 2]
```diff
@@ -1105,3 +1106,58 @@ get_rel_oid_list(void)
...
+                size += fst.st_size;
+            }
+
+            totalsize += size;
+        }
+    }
+	PG_CATCH();
+	{
+		HOLD_INTERRUPTS();
+		FlushErrorState();
```
Comment (higuoxing) [inline diskquota_utility.c:1157]: Better not to simply consume the error here.

File: Makefile [hunk 3]
```diff
@@ -5,8 +5,8 @@ MODULE_big = diskquota
 EXTENSION = diskquota
 DATA = diskquota--1.0.sql diskquota--2.0.sql diskquota--1.0--2.0.sql diskquota--2.0--1.0.sql
 SRCDIR = ./
-FILES = diskquota.c enforcement.c quotamodel.c gp_activetable.c diskquota_utility.c
-OBJS = diskquota.o enforcement.o quotamodel.o gp_activetable.o diskquota_utility.o
+FILES = diskquota.c enforcement.c quotamodel.c gp_activetable.c diskquota_utility.c diskquota_unit_test.c
+OBJS = diskquota.o enforcement.o quotamodel.o gp_activetable.o diskquota_utility.o diskquota_unit_test.o
```
Comment (higuoxing) [inline Makefile:9]: If this object file only contains helper functions for running unit tests, then I don't think it should be linked together in release builds. What about adding a switch in this `Makefile` to detect the purpose of the current build.

e.g.,

```
DEBUG=1 make ## Current build is for debugging and diskquota_unit_test.o should be linked together.
```

```
make ## Current build is for release and diskquota_unit_test.o shouldn't be linked together.
```

File: diskquota--2.0.sql [hunk 4]
```diff
@@ -117,3 +117,24 @@ LANGUAGE C VOLATILE;
 
 SELECT diskquota.diskquota_start_worker();
 DROP FUNCTION diskquota.diskquota_start_worker();
+
+-- unit test
```
Comment (higuoxing) [inline diskquota--2.0.sql:121]: I think these two SQL should be placed in `test_relation_size.sql`?

Comment (higuoxing) [inline diskquota--2.0.sql:121]: I think these two SQL should be placed in `test_relation_size.sql`?

**Review by higuoxing — COMMENTED** [review]
Couple of comments from myself. P.S. we can continuously push to the dev branch without closing the pull request.

File: diskquota--2.0.sql (hunk 4 above)
Comment (xuebinsu) [reply diskquota--2.0.sql]: Adding these two UDFs enables the user to debug Diskquota.

File: diskquota_utility.c (hunk 2 above)
Comment (xuebinsu) [reply diskquota_utility.c:1159]: Added a TODO.

File: diskquota_utility.c [hunk 5]
```diff
@@ -1190,3 +1192,80 @@ get_rel_oid_list(void)
...
+
+    rnode.node.dbNode = MyDatabaseId;
+    rnode.node.relNode = relfilenode;
+    rnode.node.spcNode = reltablespace ? reltablespace : DEFAULTTABLESPACE_OID;
+    rnode.backend = backend;
+
+    size = calculate_relation_size_all_forks(&rnode);
+
+	PG_RETURN_INT64(size);
+}
```
Comment (xuebinsu) [inline diskquota_utility.c:1271]:
```suggestion
}

```

File: sql/test_relation_size.sql [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+alter table t2 set tablespace test_spc;
+insert into t2 select generate_series(1, 10000);
+select diskquota.relation_size('t2', false);
+select pg_table_size('t2');
+
+
+drop table t1, t2;
+drop tablespace test_spc;
+\! rm -rf /tmp/test_spc
```
Comment (xuebinsu) [inline sql/test_relation_size.sql:30]:
```suggestion
\! rm -rf /tmp/test_spc

```

File: diskquota--2.0.sql [hunk 7]
```diff
@@ -127,3 +127,26 @@ LANGUAGE C VOLATILE;
...
+RETURNS bigint STRICT 
+AS $$
+SELECT sum(size)::bigint FROM (
+        SELECT diskquota.relation_size_local(reltablespace, relfilenode, is_temp) AS size 
+        FROM gp_dist_random('pg_class') WHERE oid = relation
+        UNION ALL 
+        SELECT diskquota.relation_size_local(reltablespace, relfilenode, is_temp) AS size 
+        FROM pg_class WHERE oid = relation
+) AS t
+$$ LANGUAGE SQL;
```
Comment (xuebinsu) [inline diskquota--2.0.sql:152]:
```suggestion
$$ LANGUAGE SQL;

```

**Comment by zhrt123** [comment]
+1

File: diskquota.h [hunk 8]
```diff
@@ -118,4 +118,5 @@ extern int 	SEGCOUNT;
 extern int  get_ext_major_version(void);
 extern void truncateStringInfo(StringInfo str, int nchars);
 extern List *get_rel_oid_list(void);
+extern Size diskquota_get_relation_size_by_relfilenode(RelFileNodeBackend *rnode);
```
Comment (xuebinsu) [inline diskquota.h:121]:
```suggestion

```

File: diskquota_utility.c [hunk 9]
```diff
@@ -1190,3 +1192,80 @@ get_rel_oid_list(void)
...
+static Size
+calculate_relation_size_all_forks(RelFileNodeBackend *rnode)
+{
+    int64       totalsize = 0;
+    ForkNumber  forkNum;
+    int64       size = 0;
+    char       *relationpath;
+    char        pathname[MAXPGPATH];
+    unsigned int segcount = 0;
+    PG_TRY();
```
Comment (higuoxing) [inline diskquota_utility.c:1209]: Nit: the indention is not right.

File: diskquota_utility.c [hunk 10]
```diff
@@ -1190,3 +1192,80 @@ get_rel_oid_list(void)
...
+		HOLD_INTERRUPTS();
+		FlushErrorState();
+		RESUME_INTERRUPTS();
+	}
+	PG_END_TRY();
+
+    return totalsize;
+}
+
+PG_FUNCTION_INFO_V1(relation_size_local);
```
Comment (higuoxing) [inline diskquota_utility.c:1252]: Nit: Put this declaration at the begining of this file to keep coding style consistent.

File: diskquota_utility.c [hunk 11]
```diff
@@ -1190,3 +1192,80 @@ get_rel_oid_list(void)
...
+
+Datum
+relation_size_local(PG_FUNCTION_ARGS)
+{
+    Oid reltablespace = PG_GETARG_OID(0);
```
Comment (higuoxing) [inline diskquota_utility.c:1257]: Nit: indention.

File: sql/test_relation_size.sql [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+create table t2(i int);
+insert into t2 select generate_series(1, 10000);
+select diskquota.relation_size('t2', false);
+select pg_table_size('t2');
+
+-- start_ignore
+\! mkdir /tmp/test_spc
+-- end_ignore
+DROP TABLESPACE  IF EXISTS test_spc;
+CREATE TABLESPACE test_spc LOCATION '/tmp/test_spc';
```
Comment (higuoxing) [inline sql/test_relation_size.sql:15]: Nit: keep coding style consistent.
```suggestion
drop tablespace  if exists test_spc;
create tablespace test_spc location '/tmp/test_spc';
```

File: diskquota_utility.c [hunk 13]
```diff
@@ -1190,3 +1192,80 @@ get_rel_oid_list(void)
...
+                    snprintf(pathname, MAXPGPATH, "%s",
+                            relationpath);
+                else
+                    snprintf(pathname, MAXPGPATH, "%s.%u",
+                            relationpath, segcount);
+
+                if (stat(pathname, &fst) < 0)
+                {
+                    if (errno == ENOENT)
+                        break;
```
Comment (higuoxing) [inline diskquota_utility.c:1232]: Looks that we don't handle other types of errors from `stat(2)` ? If so, we don't need to wrap this whole block using `PG_TRY()` and `PG_CATCH()`.

See: https://man7.org/linux/man-pages/man2/lstat.2.html

File: diskquota_utility.c [hunk 14]
```diff
@@ -1190,3 +1192,80 @@ get_rel_oid_list(void)
...
+{
+    Oid reltablespace = PG_GETARG_OID(0);
+	Oid relfilenode = PG_GETARG_OID(1);
+    int backend = PG_GETARG_BOOL(2) ? -2 : -1;
+    RelFileNodeBackend rnode = {0};
+    Size size = 0;
+
+    rnode.node.dbNode = MyDatabaseId;
+    rnode.node.relNode = relfilenode;
+    rnode.node.spcNode = reltablespace ? reltablespace : DEFAULTTABLESPACE_OID;
```
Comment (higuoxing) [inline diskquota_utility.c:1265]:
```suggestion
    rnode.node.spcNode = OidIsValid(reltablespace) ? reltablespace : DEFAULTTABLESPACE_OID;
```

Though C compilers can do implicit type conversion for us, I still recommend using `OidIsValid()`.

File: diskquota_utility.c [hunk 15]
```diff
@@ -1190,3 +1192,80 @@ get_rel_oid_list(void)
...
+                }
+                size += fst.st_size;
+            }
+
+            totalsize += size;
+        }
+    }
+	PG_CATCH();
+	{
+		// TODO: Record the error message to pg_log
```
Comment (higuoxing) [inline diskquota_utility.c:1242]: Nit: The comment style for C is `/* TODO: Record the error message to pg_log. */`.

File: diskquota--2.0.sql (hunk 7 above)
Comment (higuoxing) [inline diskquota--2.0.sql:152]: I think `diskquota--2.0--1.0.sql`, `diskquota--1.0--2.0.sql` should be updated for these two UDFs as well.

File: diskquota_utility.c (hunk 13 above)
Comment (xuebinsu) [reply diskquota_utility.c:1234]: I added an `else` to keep it consistent with `calculate_relation_size()`. Let's test whether we need this later.

File: diskquota--2.0.sql (hunk 7 above)
Comment (xuebinsu) [reply diskquota--2.0.sql:152]: Added a TODO item

File: diskquota_utility.c [hunk 16]
```diff
@@ -1190,3 +1194,83 @@ get_rel_oid_list(void)
 	}
 	return oidlist;
 }
+
+/*
+ * calculate size of (all forks of) a relation in transaction
+ * This function is following calculate_relation_size()
+ */
+static int64
+calculate_relation_size_all_forks(RelFileNodeBackend *rnode)
```
Comment (higuoxing) [inline diskquota_utility.c:1203]: Looks that `calculate_relation_size_all_forks` and `relation_size_local` are still indented by spaces.

**Review by higuoxing — APPROVED** [review]
+1

**Pull request merged**